### PR TITLE
Metadata: Add Getter/Setter support for indexed properties

### PR DESCRIPTION
**Pull request by smitpatel**
Part of #13610

### CODE REVIEW

File: src/EFCore/Extensions/MutableEntityTypeExtensions.cs [hunk 1]
```diff
@@ -463,6 +465,28 @@ public static IMutableProperty FindDeclaredProperty([NotNull] this IMutableEntit
...
+        /// <returns> The newly created property. </returns>
+        public static IMutableProperty AddIndexedProperty(
+            [NotNull] this IMutableEntityType entityType, [NotNull] string name, [NotNull] Type propertyType)
+        {
+            Check.NotNull(entityType, nameof(entityType));
+
+            var indexerPropertyInfo = entityType.GetIndexerPropertyInfo();
+            if (indexerPropertyInfo == null)
+            {
+                throw new InvalidOperationException(
```
Comment (AndriySvyryd) [inline src/EFCore/Extensions/MutableEntityTypeExtensions.cs]: Move the check to `GetIndexerPropertyInfo` and add `FindIndexerPropertyInfo` that doesn't throw.

File: src/EFCore/Extensions/PropertyBaseExtensions.cs [hunk 2]
```diff
@@ -82,6 +82,18 @@ public static string GetFieldName([NotNull] this IPropertyBase propertyBase)
         public static bool IsShadowProperty([NotNull] this IPropertyBase property)
             => Check.NotNull(property, nameof(property)).GetIdentifyingMemberInfo() == null;
 
+        /// <summary>
+        ///     Gets a value indicating whether this is an indexer property. An indexer property is one that is accessed through
+        ///     indexer on entity class.
```
Comment (AndriySvyryd) [inline src/EFCore/Extensions/PropertyBaseExtensions.cs]:
```suggestion
        ///     an indexer on the entity class.
```

File: src/EFCore/Extensions/PropertyBaseExtensions.cs [hunk 3]
```diff
@@ -82,6 +82,18 @@ public static string GetFieldName([NotNull] this IPropertyBase propertyBase)
...
+        /// </summary>
+        /// <param name="property"> The property to check. </param>
+        /// <returns>
+        ///     <c>True</c> if the property is an indexer property, otherwise <c>false</c>.
+        /// </returns>
+        public static bool IsIndexerProperty([NotNull] this IPropertyBase property)
+            => Check.NotNull(property, nameof(property)).GetIdentifyingMemberInfo() is PropertyInfo propertyInfo
+                && propertyInfo.IsIndexerProperty();
```
Comment (AndriySvyryd) [inline src/EFCore/Extensions/PropertyBaseExtensions.cs]:
```suggestion
                && propertyInfo == property.DeclaringType.GetIndexerPropertyInfo();
```

File: src/EFCore/Extensions/MutableEntityTypeExtensions.cs (hunk 1 above)
Comment (smitpatel) [reply src/EFCore/Extensions/MutableEntityTypeExtensions.cs]: If I move check to GetIndexerPropertyInfo then I don't have information about property name. Is that fine?

File: src/EFCore/Metadata/Internal/EntityType.cs [hunk 4]
```diff
@@ -1599,7 +1599,7 @@ private Type ValidateClrMember(string name, MemberInfo memberInfo, bool throwOnN
 
             if (name != memberInfo.GetSimpleMemberName())
             {
-                if ((memberInfo as PropertyInfo)?.IsEFIndexerProperty() != true)
+                if ((memberInfo as PropertyInfo)?.IsIndexerProperty() != true)
```
Comment (AndriySvyryd) [inline src/EFCore/Metadata/Internal/EntityType.cs]:
```suggestion
                if (memberInfo != GetIndexerPropertyInfo())
```

File: src/EFCore/Metadata/Internal/EntityType.cs [hunk 5]
```diff
@@ -2066,7 +2066,7 @@ public virtual Index RemoveIndex([NotNull] Index index)
 
             if (memberInfo != null
                 && propertyType != memberInfo.GetMemberType()
-                && (memberInfo as PropertyInfo)?.IsEFIndexerProperty() != true)
+                && (memberInfo as PropertyInfo)?.IsIndexerProperty() != true)
```
Comment (AndriySvyryd) [inline src/EFCore/Metadata/Internal/EntityType.cs]: same as above

File: src/EFCore/Metadata/Internal/InternalEntityTypeBuilder.cs [hunk 6]
```diff
@@ -682,7 +682,7 @@ public virtual bool CanAddProperty([NotNull] Type propertyType, [NotNull] string
             var memberInfo = Metadata.ClrType.GetMembersInHierarchy(name).FirstOrDefault();
             if (memberInfo != null
                 && propertyType != memberInfo.GetMemberType()
-                && (memberInfo as PropertyInfo)?.IsEFIndexerProperty() != true
+                && (memberInfo as PropertyInfo)?.IsIndexerProperty() != true
```
Comment (AndriySvyryd) [inline src/EFCore/Metadata/Internal/InternalEntityTypeBuilder.cs]: ditto

File: src/EFCore/Metadata/Internal/TypeBase.cs [hunk 7]
```diff
@@ -168,6 +169,29 @@ public virtual void UpdateConfigurationSource(ConfigurationSource configurationS
...
+        ///     doing so can result in application failures when updating to a new Entity Framework Core release.
+        /// </summary>
+        public virtual PropertyInfo GetIndexerPropertyInfo()
+        {
+            if (ClrType == null)
+            {
+                return null;
+            }
+
+            if (_indexerPropertyInfo == null)
```
Comment (AndriySvyryd) [inline src/EFCore/Metadata/Internal/TypeBase.cs]: Use a sentinel value or a bool field to indicate that `_indexerPropertyInfo` wasn't initialized as opposed to no indexer on the type.

File: src/EFCore/Metadata/Internal/TypeBase.cs [hunk 8]
```diff
@@ -198,6 +222,7 @@ public virtual void ClearCaches()
         {
             _runtimeProperties = null;
             _runtimeFields = null;
+            _indexerPropertyInfo = null;
```
Comment (AndriySvyryd) [inline src/EFCore/Metadata/Internal/TypeBase.cs]: This method isn't even used anywhere, just remove it.

File: src/EFCore/Metadata/Internal/TypeBaseExtensions.cs [hunk 9]
```diff
@@ -53,17 +62,9 @@ public static bool HasClrType([NotNull] this ITypeBase type)
...
-                return property;
-            }
-
-            if (type.GetRuntimeFields().TryGetValue(name, out var field))
-            {
-                return field;
-            }
-
-            return null;
+            return type.GetRuntimeProperties().TryGetValue(name, out var property)
```
Comment (AndriySvyryd) [inline src/EFCore/Metadata/Internal/TypeBaseExtensions.cs]: Expression body + formatting.

File: src/EFCore/Extensions/MutableEntityTypeExtensions.cs (hunk 1 above)
Comment (smitpatel) [reply src/EFCore/Extensions/MutableEntityTypeExtensions.cs]: Discussed in person decided to keep it here only for better exception message.

File: src/EFCore/Metadata/Internal/TypeBase.cs [hunk 10]
```diff
@@ -174,32 +176,43 @@ public virtual void UpdateConfigurationSource(ConfigurationSource configurationS
...
+        public virtual PropertyInfo FindIndexerPropertyInfo()
+        {
+            if (ClrType == null)
+            {
+                return null;
+            }
+
+            if (!_indexerPropertyInitialized)
+            {
+                var indexerPropertyInfo = GetRuntimeProperties().Values.FirstOrDefault(pi => pi.IsIndexerProperty());
```
Comment (AndriySvyryd) [inline src/EFCore/Metadata/Internal/TypeBase.cs]:
```suggestion
                var indexerPropertyInfo = GetRuntimeProperties().Values.FirstOrDefault(PropertyInfoExtensions.IsIndexerProperty);
```

Comment (smitpatel) [reply src/EFCore/Metadata/Internal/TypeBase.cs]: Updated in 3rd PR in this series.

**Comment by smitpatel** [comment]
Updated.

**Pull request merged**